File: LeetCode/104-maximum-depth-of-binary-tree/Solution.java
import java.util.Stack;

public class Solution {
  /**
   * Finds the maximum depth of a binary tree using an iterative approach.
   *
   * @param root The root node of the binary tree.
   * @return The maximum depth of the binary tree.
   */
  public int maxDepth(TreeNode root) {
    if (root == null) {
      return 0;
    }

    Stack<Pair<TreeNode, Integer>> stack = new Stack<>();
    stack.push(new Pair<>(root, 1));
    int maxDepth = 0;

    while (!stack.isEmpty()) {
      Pair<TreeNode, Integer> current = stack.pop();
      TreeNode currentNode = current.getKey();
      int currentDepth = current.getValue();

      if (currentNode != null) {
        maxDepth = Math.max(maxDepth, currentDepth);
        stack.push(new Pair<>(currentNode.left, currentDepth + 1));
        stack.push(new Pair<>(currentNode.right, currentDepth + 1));
      }
    }

    return maxDepth;
  }
}

// Auxiliary class to store tree node and its depth
class Pair<K, V> {
  private final K key;
  private final V value;

  public Pair(K key, V value) {
    this.key = key;
    this.value = value;
  }

  public K getKey() {
    return key;
  }

  public V getValue() {
    return value;
  }
}

File: LeetCode/104-maximum-depth-of-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * Finds the maximum depth of a binary tree using an iterative approach.
 *
 * @param root The root node of the binary tree.
 * @returns The maximum depth of the binary tree.
 */
function maxDepth(root: TreeNode | null): number {
  if (root === null) {
    return 0;
  }

  const stack: Array<{ node: TreeNode | null; depth: number }> = [
    { node: root, depth: 1 },
  ];
  let maxDepth = 0;

  while (stack.length > 0) {
    const { node, depth } = stack.pop()!;
    if (node) {
      maxDepth = Math.max(maxDepth, depth);
      if (node.left !== null) {
        stack.push({ node: node.left, depth: depth + 1 });
      }
      if (node.right !== null) {
        stack.push({ node: node.right, depth: depth + 1 });
      }
    }
  }

  return maxDepth;
};
